### ISSUE TITLE
GoCD server crashing under load

### ISSUE THREAD

**Issue by Confused82** [gocd/gocd#5588]
Hi, 

I'm not sure if this is a GoCD problem or something else, but I'd be very grateful for any advice. I'm running Go version 18.11.0 on an m4.xlarge in AWS with 4 internal agents (i.e. not dockerised). Most of the jobs involve running maven. Under load ( ie all 4 agents working constantly for 30 mins or so) the server falls over - although the instance is still 'up', the go-server has stopped, and the instance can't be SSHd on to. Rebooting/restarting the instance does not work - the newly restarted instance turns itself off again. As a consequence I can't get any logs off the instance. The only hint I have are the failed pipelines that were running when the server crashed, which sometimes mention a lack of memory, although monitoring the instance memory up to the crash does not show any problems. 

Any thoughts? Thanks so much for your help.

**Comment by arvindsv** [comment]
@Confused82 I'd look at the output of `dmesg` (kernel messages) to see if the oomkiller killed a java process.

Don't know why the instance cannot be SSH'ed into. That shows (to me) that it is something beyond what the GoCD server is doing. Actually, one thing that can be happening is: You could be running out of IOPS. Every EC2 node has a set number of IOPS that it can use, before the node is suspended for some time. These IOPS are generated in some time and then, the node continues working. I think @barrowkwan knows this well, based on a support ticket with AWS.

tl; dr - Check the iops in the mnitoring section of that EC2 node and check other limits, such as network access, etc.

**Comment by arvindsv** [comment]
"These IOPS are generated in some time" => "You get some iops per duration". So, after being suspended, it'll get back - and if it uses too much, gets suspended again.

**Comment by arvindsv** [comment]
Could also be CPU credits, etc. Basically anything that goes "over", will cause your node to slow down to a crawl or be suspended for a while.

**Comment by Confused82** [comment]
Thanks for the ideas, I'm trying out ways to catch error messages, as after the services like go-server is terminated, there is no way to log on to the box any more to check the kernel messages. I've tried running vmstat as the box is crashing (see below), and it does look like the instance is running out of memory, with several java procsses taking it all. All our pipeline steps are dockerised, so I'm guessing it's the go-server itself. For the moment I'm thinking about settings JVM memory limits and see if that helps, and push kernel messages to EFS so I can catch them later.

Thanks again for your help!

procs -----------memory---------- ---swap-- -----io---- --system-- -----cpu-----
 r  b   swpd   free   buff  cache   si   so    bi    bo   in   cs us sy id wa st
 0  0      0   4114    231   8264    0    0    20   165  403  516  4  1 94  1  0	
 0  0      0   4114    231   8264    0    0     0    11  680  964  0  0 100  0  0	
 0  0      0   4111    231   8264    0    0     0    13  926 1201  1  0 99  0  0	
 0  0      0   4107    232   8265    0    0     1   183 1158 1423  2  0 97  0  0	
 0  0      0   4106    232   8266    0    0     0    45  860 1138  1  0 99  0  0	
 0  0      0   4110    232   8265    0    0     0    16  796 1075  1  0 99  0  0	
 0  0      0   4110    232   8265    0    0     0    11  897 1181  1  0 99  0  0	
 0  0      0   4110    232   8266    0    0     0    42  685  964  0  0 100  0  0	
 0  0      0   4168    232   8225    0    0     0    12 1003 1257  2  0 98  0  0	
 0  1      0   3299    234   8505    0    0     0  3615 8153 10460 41  4 48  7  0	
 0  0      0   3138    240   9158    0    0     0 21102 14325 17584 48  8 33 10  0	
 6  0      0   2636    247   9347    0    0     0  4260 8312 12862 55  6 39  0  0	
 3  0      0   2597    250   9645    0    0     0  6038 7308 8459 42  5 41 12  0	
 1  0      0   1677    251   9486    0    0     0  2707 8157 11857 43  5 50  2  0	
 0  2      0   3499    258   8770    0    0    22  8860 11877 16889 60  8 23 10  0	
 0  0      0   3571    258   8674    0    0     0  1301 3772 4226  1  1 66 32  0	
 3  0      0   2609    266   9515    0    0     0   473 12172 13630 12  4 82  1  0	
 2  0      0   1691    272   9508    0    0     0 21314 9656 14843 52  6 38  5  0	
 0  1      0   2326    277   9795    0    0     0  5579 10138 14027 42  5 41 12  0	
 0  0      0   2083    278   9752    0    0     0   493 4885 5914 10  2 67 21  0	
 2  0      0   1698    281  10030    0    0     0  5038 9636 13277 55  6 36  3  0	
procs -----------memory---------- ---swap-- -----io---- --system-- -----cpu-----
 r  b   swpd   free   buff  cache   si   so    bi    bo   in   cs us sy id wa st
 3  1      0   1292    290  10672    0    0     0 18144 12551 14920 33  6 46 15  0	
 3  0      0    453    297  10848    0    0     0  3223 8554 10576 33  4 55  9  0

**Comment by arvindsv** [comment]
@Confused82 Cool. I'd be suspecting the maven processes, rather than the GoCD server. :) But, I'll wait to hear from you. If it is the GoCD server causing it, I'm all ears.

**Comment by arvindsv** [comment]
Tentatively closing this, till we hear back.
